com.unity.cinemachine/Runtime/Debug/CinemachineDebug.cs

```diff
@@ -0,0 +1,138 @@
+        /// <param name="text">Sample text, for determining rectangle size</param>
+        /// <param name="style">What style will be used to draw, used here for
+        /// determining rect size</param>
+        /// <returns>An area on the game screen large enough to print the text
+        /// in the style indicated</returns>
+        public static Rect GetScreenPos(Camera camera, string text, GUIStyle style)
+        {
+            var size = style.CalcSize(new GUIContent(text));
+            var pos = Vector2.zero;
+            if (camera != null)
+            {
+                var r = camera.rect; 
+                pos.x += r.x * Screen.width;
+                pos.y += (1 - r.y - r.height) * Screen.height;
+            }
+            return new Rect(pos, size);
+        }
+#endif
+
+        
+#if UNITY_EDITOR
+        const string k_DebugUIName = "CinemachineDebugUI";
+        static Dictionary<Camera, VisualElement> s_CameraViewRectContainers = new();
+        static GameObject s_UIDocumentHolder;
+        static UIDocument s_UIDocument;
+        public static VisualElement CreateRuntimeUIContainer(Camera outputCamera)
+        {
+            if (s_UIDocumentHolder == null)
+            {
+                s_UIDocumentHolder = GameObject.Find(k_DebugUIName);
+                if (s_UIDocumentHolder == null)
+                {
+                    s_UIDocumentHolder = new GameObject(k_DebugUIName)
+                    {
+                        hideFlags = HideFlags.NotEditable | HideFlags.HideInHierarchy,
+                        tag = "EditorOnly" // TODO: we may want runtime too
+                    };
+                }
+            }
+            if (!s_UIDocumentHolder.TryGetComponent(out s_UIDocument))
+            {
+                s_UIDocument = s_UIDocumentHolder.AddComponent<UIDocument>();
+                
+                const string path = "Packages/com.unity.cinemachine/Runtime/UI/";
+                s_UIDocument.panelSettings = AssetDatabase.LoadAssetAtPath<PanelSettings>(path + "CinemachinePanelSettings.asset"); // TODO: find a runtime compatible way
+            }
+
+            if (s_CameraViewRectContainers.ContainsKey(outputCamera))
+                return s_CameraViewRectContainers[outputCamera];
+
+            var viewportContainer = new VisualElement
+            {
+                name = "CinemachineDebugUI_ViewportContainer",
+                style =
+                {
+                    position = new StyleEnum<Position>(Position.Absolute),
+                    flexWrap = new StyleEnum<Wrap>(Wrap.Wrap),
+                }
+            };
+            // need to use delayCall, because rootVisualElement may not be built at this point yet
+            EditorApplication.delayCall += () => s_UIDocument.rootVisualElement.Add(viewportContainer);
+            viewportContainer.schedule.Execute(() => PositionWithinCameraView(viewportContainer, outputCamera)).Every(0); // TODO: can we do better?
```

Comment:
This is going to be running whether or not there is debug text displayed.  I'm sure we can improve that. We should be updating the containers only if they're actually doing any work.